framework/src/base/MooseApp.C

```diff
@@ -610,9 +618,370 @@ MooseApp::MooseApp(InputParameters parameters)
+                             false,
+                             "Trapping floating point exceptions is not enabled (enable them using "
+                             "the --trap-fpe option or by running a debug mode executable).");
+
+  {
+    const auto doc = "LibTorch machine learning and parallel tensor algebra library";
+#ifdef LIBTORCH_ENABLED
+    haveCapability("libtorch", doc);
+#else
+    missingCapability("libtorch",
+                      doc,
+                      "Check "
+                      "https://mooseframework.inl.gov/moose/getting_started/installation/"
+                      "install_libtorch.html for "
+                      "instructions on how to configure and build moose with libTorch.");
+#endif
+  }
+
+  {
+    const auto doc = "gperftools code performance analysis and profiling library";
+#ifdef HAVE_GPERFTOOLS
+    haveCapability("gperftools", doc);
+#else
+    missingCapability("gperftools",
+                      doc,
+                      "Check https://mooseframework.inl.gov/application_development/profiling.html "
+                      "for instructions on profiling MOOSE based applications.");
+#endif
+  }
+
+  {
+    const auto doc = "libPNG portable network graphics format library";
+#ifdef MOOSE_HAVE_LIBPNG
+    haveCapability("libpng", doc);
+#else
+    missingCapability("libpng",
+                      doc,
+                      "Install libpng through conda or your distribution and check that it gets "
+                      "detected through pkg-config, then reconfigure and rebuild MOOSE.");
+#endif
+  }
+
+  {
+    const auto doc = "NVIDIA GPU parallel computing platform";
+#ifdef CUDA_SUPPORTED
+    haveCapability("cuda", doc);
+#else
+    missingCapability("cuda", doc, "Add the CUDA bin directory to your path and rebuild PETSc.");
+#endif
+  }
+
+  Moose::Capabilities::add(
+      "ad_size",
+      MOOSE_AD_MAX_DOFS_PER_ELEM,
+      "MOOSE was configured and built with a dual number backing store size of " +
+          Moose::stringify(MOOSE_AD_MAX_DOFS_PER_ELEM) +
+          ". Complex simulations with many variables or contact problems may require larger "
+          "values. Reconfigure MOOSE with the --with-derivative-size=<n> option in the root of the "
+          "repository.");
+  {
+    const std::string method = QUOTE(METHOD);
+    Moose::Capabilities::add(
+        "method", method, "The executable was built with METHOD=\"" + method + "\"");
+  }
+
+  {
+    const auto doc = "Workbench Analysis Sequence Processors (WASP) parser library";
+#ifdef WASP_ENABLED
+    haveCapability("wasp", doc);
+#else
+    missingCapability("wasp", doc, " Unable to parse input files.");
+#endif
+  }
+
+  {
+    const std::string version = QUOTE(LIBMESH_DETECTED_PETSC_VERSION_MAJOR) "." QUOTE(
+        LIBMESH_DETECTED_PETSC_VERSION_MINOR) "." QUOTE(LIBMESH_DETECTED_PETSC_VERSION_SUBMINOR);
+    Moose::Capabilities::add("petsc", version, "Using PETSc version " + version + ".");
+  }
```

Comment:
what if we don't detect a petsc version (non-release hash)

Reply:
Hmm, yeah, what then? We cannot know if a random hash is greater or less than a specific version (at least not without inspecting the PETSc repo...)

Reply:
@loganharbour we should discuss this. I cannot remember the last time we actually tagged a PETSc non-release. If that will be required we could add an additional capability that is set to `false` if no non-release is used, and to the hash (string) if it is, and conversely we could set `petsc` to `false` if a non-release hash is used.

Reply:
Or set `petsc` to `true`. Any comparison with a bool `true` capability succeeds and any comparison with a bool `false` capability fails. That can be used to engineer "always match" versions or "always fail" versions.